mmdet/core/bbox/assigners/hungarian_assigner.py

```diff
@@ -116,14 +119,26 @@ def assign(self,
         # NLL is used, we approximate it in 1 - cls_score[gt_label].
         # The 1 is a constant that doesn't change the matching,
         # so it can be ommitted.
-        cls_score = cls_pred.softmax(-1)
-        cls_cost = -cls_score[:, gt_labels]  # [num_bboxes, num_gt]
+        # MODYFY add focal loss
+        if self.focal_loss:
+            alpha = self.focal_loss['alpha']
+            gamma = self.focal_loss['gamma']
+            eps = 1e-12
+            neg_loss = -(1 - cls_pred + eps).log() * (
+                        1 - alpha) * cls_pred.pow(gamma)
+            pos_loss = -(cls_pred + eps).log() * alpha * (1 - cls_pred).pow(
+                gamma)
+            cls_cost = pos_loss[:, gt_labels] - neg_loss[:, gt_labels]
+        else:
+            cls_score = cls_pred.softmax(-1)
+            cls_cost = -cls_score[:, gt_labels]  # [num_bboxes, num_gt]
 
         # regression L1 cost
         img_h, img_w, _ = img_meta['img_shape']
         factor = torch.Tensor([img_w, img_h, img_w,
                                img_h]).unsqueeze(0).to(gt_bboxes.device)
         gt_bboxes_normalized = gt_bboxes / factor
+        # MODYFY: xyxy p1-norm in sparse-RCNN CODE which need to review
```

Comment:
We may also use build loss for bbox distance.

Reply:
updated